guide.md

```diff
@@ -0,0 +1,324 @@
+
+Bazel is valued for its hermetic and reproducible builds, but this example
+already shows that that isn't completely accurate: we haven't told Bazel what
+C++ compiler to use or where to find it and all the auxiliary tools it relies
+on (linker, assembler, and the like). It simply looks for appropriately-named
+executables in the PATH and sets up a C/C++ compiler toolchain called
+`@local_config_cc` that "should work". 
+
+This is problematic, and not just because of academic concerns regarding the
+precise meaning of "hermeticity": build products output by different versions
+of the same compiler can vary greatly, let alone between different compilers.
+Bazel, by default, chooses GCC on a Linux machine and Clang/LLVM on a macOS
+machine, which can mean that the same bit of hot code, say, vectorizes
+correctly on one platform but not on the other due to some difference between
+the two compilers, resulting in wildly different performance characteristics.
+
+There are Bazel-native ways to mitigate these issues, but they fall short of true
+hermeticity: for instance, `--incompatible_strict_action_env` runs actions in
+an environment that does not inherit a `PATH` from the shell environment, but
+instead [sets it to][path-link] the static value `/bin:/usr/bin:/usr/local/bin`.
+This means that as long as your compiler tools are in one of those paths, which
+they almost certainly are on \*nix systems, nothing changes.
+
+# How rules_nixpkgs solves the issue
+
+There are two classes of workaround: one is to give up trying to make Bazel not
+depend on the global environment, and instead to control that global
+environment ourselves so that the implicit dependencies end up being the same
+every time. To do so, we run all builds in sandboxes or VMs that all share the
+exact same configuration, for instance, in identical Docker containers.
+
+The other solution is to eliminate the dependency on the global environment by
+defining your own compiler toolchain, providing paths to executable binaries that you
+provision and version yourself. Bazel can then be instructed to use that
+instead of the default `@local_config_cc`. This can be done by checking all the
+relevant executables (`cc`, `ld`, `as`, ...) into your repository yourself, as
+is done within Google for instance. 
+
+In both cases, the build depends on non-code elements (Docker containers,
+executable binaries) for which versioning and maintenance is more complex than
+for ordinary code. Extra steps must now be followed when these environments
+have to be changed or when new releases are cut. Either these configuration
+files have to be executed at runtime, necessitating long setup processes for
+each executor, or one caches the build products. In the latter case, large
+binary blobs have to be managed and moved from machine to machine, often in the
+hundreds of gigabytes for Docker images.
+
+The r_n approach is a variation on the second, where instead of managing system
```

Comment:
I'd characterize it as a different approach rather than a variation. The reason is that it enables fine grained dependency tracking on system dependencies and toolchains. E.g. if we only add a system library dependency imported through rules_nixpkgs, it won't affect the CC toolchain and we won't have to invalidate previously built CC targets.

One reason I'd make that distinction is that a variation of the Docker approach is indeed possible with Nix. We could use Nix Docker tools to build a Docker image with the build environment in a reproducible way. However, this still has the issue of being coarse grained, since Bazel doesn't track precisely which parts of the image the build depends on.

Reply:
Good point, I've clarified this.